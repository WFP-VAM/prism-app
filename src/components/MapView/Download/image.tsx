import {
  Button,
  createStyles,
  Dialog,
  DialogActions,
  DialogContent,
  DialogTitle,
  Theme,
  WithStyles,
  withStyles,
} from '@material-ui/core';
import { jsPDF } from 'jspdf';
import React from 'react';
import { useSafeTranslation } from '../../../i18n';

function DownloadImage({
  classes,
  open,
  setOpen,
  previewRef,
  handleClose,
}: DownloadImageProps) {
  const { t } = useSafeTranslation();

  const download = (format: string) => {
    const ext = format === 'pdf' ? 'png' : format;
    const canvas = previewRef.current;
    if (canvas) {
      const file = canvas.toDataURL(`image/${ext}`);
      if (format === 'pdf') {
        // eslint-disable-next-line new-cap
        const pdf = new jsPDF({
          orientation: 'landscape',
        });
        const imgProps = pdf.getImageProperties(file);
        const pdfWidth = pdf.internal.pageSize.getWidth();
        const pdfHeight = (imgProps.height * pdfWidth) / imgProps.width;
        pdf.addImage(file, 'PNG', 0, 0, pdfWidth, pdfHeight);
        pdf.save('map.pdf');
      } else {
        const link = document.createElement('a');
        link.setAttribute('href', file);
        link.setAttribute('download', `map.${ext}`);
        link.click();
      }
      setOpen(false);
      handleClose();
    }
  };

  return (
    <Dialog
      maxWidth="xl"
      open={open}
      keepMounted
      onClose={() => setOpen(false)}
      aria-labelledby="dialog-preview"
    >
      <DialogTitle className={classes.title} id="dialog-preview">
        {t('Map Preview')}
      </DialogTitle>
      <DialogContent>
        <canvas ref={previewRef} />
      </DialogContent>
      <DialogActions>
        <Button onClick={() => setOpen(false)} color="primary">
          {t('Cancel')}
        </Button>
        <Button
          variant="contained"
          onClick={() => download('png')}
          color="primary"
        >
          {t('Download PNG')}
        </Button>
        <Button
          variant="contained"
          onClick={() => download('jpeg')}
          color="primary"
        >
          {t('Download JPEG')}
        </Button>
        <Button
          variant="contained"
          onClick={() => download('pdf')}
          color="primary"
        >
          {t('Download PDF')}
        </Button>
      </DialogActions>
    </Dialog>
  );
}

const styles = (theme: Theme) =>
  createStyles({
    title: {
      color: theme.palette.text.secondary,
    },
  });

export interface DownloadImageProps extends WithStyles<typeof styles> {
  open: boolean;
  setOpen: React.Dispatch<React.SetStateAction<boolean>>;
  previewRef: React.RefObject<HTMLCanvasElement>;
  handleClose: () => void;
}

export default withStyles(styles)(DownloadImage);
