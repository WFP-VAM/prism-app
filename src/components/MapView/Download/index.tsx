import React, { useRef, useState } from 'react';
import moment from 'moment';
import {
  Button,
  createStyles,
  Dialog,
  DialogActions,
  DialogContent,
  DialogTitle,
  Grid,
  Hidden,
  ListItemIcon,
  ListItemText,
  MenuItem,
  Theme,
  Typography,
  WithStyles,
  withStyles,
} from '@material-ui/core';
import Menu, { MenuProps } from '@material-ui/core/Menu';
import {
  CloudDownload,
  ArrowDropDown,
  Image,
  Description,
} from '@material-ui/icons';
import { jsPDF } from 'jspdf';
import { useSelector } from 'react-redux';
import {
  mapSelector,
  layersSelector,
  dateRangeSelector,
} from '../../../context/mapStateSlice/selectors';

type ExtraDownloadRequest = {
  url: string;
  date: number | undefined;
};

const ExportMenu = withStyles((theme: Theme) => ({
  paper: {
    border: '1px solid #d3d4d5',
    backgroundColor: theme.palette.primary.main,
  },
}))((props: MenuProps) => (
  <Menu
    elevation={0}
    getContentAnchorEl={null}
    anchorOrigin={{
      vertical: 'bottom',
      horizontal: 'center',
    }}
    transformOrigin={{
      vertical: 'top',
      horizontal: 'center',
    }}
    {...props}
  />
));

const ExportMenuItem = withStyles((theme: Theme) => ({
  root: {
    color: theme.palette.common.white,
  },
}))(MenuItem);

function Download({ classes }: DownloadProps) {
  const [anchorEl, setAnchorEl] = useState<null | HTMLElement>(null);
  const [open, setOpen] = useState(false);
  const selectedMap = useSelector(mapSelector);
  const previewRef = useRef<HTMLCanvasElement>(null);

  const handleClick = (event: React.MouseEvent<HTMLElement>) => {
    setAnchorEl(event.currentTarget);
  };

  const handleClose = () => {
    setAnchorEl(null);
  };

  const openModal = () => {
    if (selectedMap) {
      const activeLayers = selectedMap.getCanvas();
      const canvas = previewRef.current;
      if (canvas) {
        canvas.setAttribute('width', activeLayers.width.toString());
        canvas.setAttribute('height', activeLayers.height.toString());
        const context = canvas.getContext('2d');
        if (context) {
          context.drawImage(activeLayers, 0, 0);
        }
      }
      setOpen(true);
    }
    handleClose();
  };

  const download = (format: string) => {
    const ext = format === 'pdf' ? 'png' : format;
    const canvas = previewRef.current;
    if (canvas) {
      const file = canvas.toDataURL(`image/${ext}`);
      if (format === 'pdf') {
        // eslint-disable-next-line new-cap
        const pdf = new jsPDF({
          orientation: 'landscape',
        });
        const imgProps = pdf.getImageProperties(file);
        const pdfWidth = pdf.internal.pageSize.getWidth();
        const pdfHeight = (imgProps.height * pdfWidth) / imgProps.width;
        pdf.addImage(file, 'PNG', 0, 0, pdfWidth, pdfHeight);
        pdf.save('map.pdf');
      } else {
        const link = document.createElement('a');
        link.setAttribute('href', file);
        link.setAttribute('download', `map.${ext}`);
        link.click();
      }
      setOpen(false);
      handleClose();
    }
  };

  function downloadExtra(request: ExtraDownloadRequest) {
    const url = new URL(request.url);
    if (request.date) {
      url.searchParams.append(
        'date',
        moment(request.date).format('YYYY-MM-DD'),
      );
    }
    window.open(url.toString());
  }

  const layers = useSelector(layersSelector);
  // first layer is always boundary layer
  const lastLayer = layers.length > 1 ? layers[layers.length - 1] : null;
  const extraDownloads =
    lastLayer && lastLayer.downloads ? lastLayer.downloads : [];
  const selectedDate = useSelector(dateRangeSelector);

  return (
    <Grid item>
      <Button variant="contained" color="primary" onClick={handleClick}>
        <CloudDownload fontSize="small" />
        <Hidden smDown>
          <Typography className={classes.label} variant="body2">
            Export
          </Typography>
        </Hidden>
        <ArrowDropDown fontSize="small" />
      </Button>
      <ExportMenu
        id="export-menu"
        anchorEl={anchorEl}
        keepMounted
        open={Boolean(anchorEl)}
        onClose={handleClose}
      >
        <ExportMenuItem onClick={openModal}>
          <ListItemIcon>
            <Image fontSize="small" style={{ color: 'white' }} />
          </ListItemIcon>
          <ListItemText primary="IMAGE" />
        </ExportMenuItem>
        {extraDownloads.map(ed => (
          <ExportMenuItem
            key={ed.label}
            onClick={() => {
              downloadExtra({
                date: selectedDate.startDate,
                url: ed.url,
              });
            }}
          >
            <ListItemIcon>
              <Description fontSize="small" style={{ color: 'white' }} />
            </ListItemIcon>
            <ListItemText primary={ed.label} />
          </ExportMenuItem>
        ))}
      </ExportMenu>
      <Dialog
        maxWidth="xl"
        open={open}
        keepMounted
        onClose={() => setOpen(false)}
        aria-labelledby="dialog-preview"
      >
        <DialogTitle className={classes.title} id="dialog-preview">
          Map Preview
        </DialogTitle>
        <DialogContent>
          <canvas ref={previewRef} />
        </DialogContent>
        <DialogActions>
          <Button onClick={() => setOpen(false)} color="primary">
            Cancel
          </Button>
          <Button
            variant="contained"
            onClick={() => download('png')}
            color="primary"
          >
            Download PNG
          </Button>
          <Button
            variant="contained"
            onClick={() => download('jpeg')}
            color="primary"
          >
            Download JPEG
          </Button>
          <Button
            variant="contained"
            onClick={() => download('pdf')}
            color="primary"
          >
            Download PDF
          </Button>
        </DialogActions>
      </Dialog>
    </Grid>
  );
}

const styles = (theme: Theme) =>
  createStyles({
    label: {
      marginLeft: '10px',
    },
    title: {
      color: theme.palette.text.secondary,
    },
  });

export interface DownloadProps extends WithStyles<typeof styles> {}

export default withStyles(styles)(Download);
