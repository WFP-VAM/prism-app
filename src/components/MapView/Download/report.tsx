import React from 'react';
import {
  Button,
  createStyles,
  Dialog,
  DialogActions,
  DialogContent,
  DialogTitle,
  WithStyles,
  withStyles,
} from '@material-ui/core';
import { jsPDF } from 'jspdf';
import { Page, Document, pdfjs } from 'react-pdf';
import { useSafeTranslation } from '../../../i18n';

// eslint-disable-next-line fp/no-mutation
pdfjs.GlobalWorkerOptions.workerSrc = `//unpkg.com/pdfjs-dist@${pdfjs.version}/build/pdf.worker.min.js`;

function Report({
  // eslint-disable-next-line no-unused-vars
  classes,
  open,
  setOpen,
  // eslint-disable-next-line no-unused-vars
  previewRef,
  handleClose,
}: ReportProps) {
  const { t } = useSafeTranslation();
  const [preview, setPreview] = React.useState<any>();

  function blobToBase64(blob: Blob): Promise<string> {
    // eslint-disable-next-line no-unused-vars
    return new Promise((resolve: (value: string) => void, _) => {
      const reader = new FileReader();
      // eslint-disable-next-line fp/no-mutation
      reader.onloadend = () => resolve(reader.result as string);
      reader.readAsDataURL(blob);
    });
  }

  async function updatePreview(blobPDF: any) {
    const result = await blobToBase64(blobPDF);
    setPreview(result);
  }

  function removeTextLayerOffset() {
    const textLayers = document.querySelectorAll(
      '.react-pdf__Page__textContent',
    );
    textLayers.forEach(layer => {
      const { style } = layer as any;
      // eslint-disable-next-line fp/no-mutation
      style.top = '0';
      // eslint-disable-next-line fp/no-mutation
      style.left = '0';
      // eslint-disable-next-line fp/no-mutation
      style.transform = '';
    });
  }

  React.useEffect(() => {
    // eslint-disable-next-line new-cap
    const pdf = new jsPDF();

    pdf.text('Hello world!', 10, 10);

    const blobPDF = new Blob([pdf.output('blob')], { type: 'application/pdf' });
    updatePreview(blobPDF);

    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, []);

  return (
    <Dialog
      open={open}
      keepMounted
      onClose={() => setOpen(false)}
      maxWidth={false}
    >
      <DialogTitle>{t('Report Title')}</DialogTitle>
      <DialogContent>
        <div>
          <Document file={preview}>
            <Page pageNumber={1} onLoadSuccess={removeTextLayerOffset} />
          </Document>
        </div>
      </DialogContent>
      <DialogActions>
        <Button
          onClick={() => {
            setOpen(false);
            handleClose();
          }}
          color="primary"
        >
          {t('Cancel')}
        </Button>
      </DialogActions>
    </Dialog>
  );
}

const styles = () => createStyles({});

export interface ReportProps extends WithStyles<typeof styles> {
  open: boolean;
  setOpen: React.Dispatch<React.SetStateAction<boolean>>;
  previewRef: React.RefObject<HTMLCanvasElement>;
  handleClose: () => void;
}

export default withStyles(styles)(Report);
