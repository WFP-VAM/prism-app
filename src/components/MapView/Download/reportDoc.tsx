import React from 'react';
import {
  Page,
  Document,
  StyleSheet,
  View,
  Text,
  Image,
} from '@react-pdf/renderer';
import { Theme } from '@material-ui/core';
import { chunk } from 'lodash';
import { Style } from '@react-pdf/types';
import {
  TableData,
  TableRowType,
} from '../../../context/analysisResultStateSlice';
import { getLegendItemLabel, ReportType } from '../utils';
import { LegendDefinition } from '../../../config/types';
import { getTableCellVal, TFunction } from '../../../utils/data-utils';

// This numbers depends on table header and row height
const MAX_TABLE_ROWS_PER_PAGE = 29;
const FIRST_PAGE_TABLE_ROWS = 9;

const makeStyles = (theme: Theme) =>
  StyleSheet.create({
    page: {
      flexDirection: 'column',
      backgroundColor: '#FFF',
      paddingBottom: 25,
      paddingTop: '1vh',
      fontFamily: 'Roboto',
    },
    section: {
      width: '96vw',
      marginLeft: '2vw',
      marginRight: '2vw',
      marginBottom: 10,
    },
    title: {
      fontSize: theme.pdf?.fontSizes.large,
      fontWeight: 500,
    },
    subText: {
      fontSize: theme.pdf?.fontSizes.medium,
    },
    mapImage: {
      width: '100%',
      maxHeight: 284,
    },
    legendsContainer: {
      display: 'flex',
      flexDirection: 'row',
      flexWrap: 'wrap',
      justifyContent: 'flex-start',
      backgroundColor: theme.pdf?.legendsBackgroundColor,
    },
    legend: {
      height: 130,
      display: 'flex',
      flexDirection: 'column',
      marginRight: 10,
      padding: 10,
    },
    legendTittle: {
      fontSize: theme.pdf?.fontSizes.large,
      paddingBottom: 10,
      fontWeight: 500,
    },
    legendContentsWrapper: {
      display: 'flex',
      flexDirection: 'column',
      flexWrap: 'wrap',
      height: '100%',
    },
    legendContent: {
      display: 'flex',
      flexDirection: 'row',
      alignItems: 'center',
      fontSize: theme.pdf?.fontSizes.large,
      paddingTop: 5,
      paddingRight: 5,
    },
    legendText: {
      paddingLeft: 10,
    },
    dash: {
      width: 16,
      height: 2,
    },
    borderedBox: {
      width: 12,
      height: 12,
      border: 2,
      borderRadius: 3,
    },
    box: {
      width: 10,
      height: 10,
    },
    tableName: {
      padding: 6,
      fontWeight: 500,
    },
    tableCell: {
      padding: 8,
      fontSize: theme.pdf?.fontSizes.medium,
    },
    tableRow: {
      display: 'flex',
      flexDirection: 'row',
      lineHeight: 1,
    },
    tableHead: {
      borderBottom: 1,
      borderBottomColor: theme.pdf?.table?.borderColor,
      backgroundColor: theme.pdf?.table?.darkRowColor,
      fontWeight: 500,
    },
    tableFooter: {
      borderTop: 1,
      borderTopColor: theme.pdf?.table?.borderColor,
      fontWeight: 700,
    },
    footer: {
      position: 'absolute',
      fontSize: theme.pdf?.fontSizes.medium,
      bottom: 10,
      left: 10,
      right: 0,
      textAlign: 'left',
    },
  });

interface PDFLegendDefinition {
  value: string | number;
  style: Style | Style[];
}

interface LegendProps {
  theme: Theme;
  title: string;
  definition: PDFLegendDefinition[];
}

const Legend = ({ theme, title, definition }: LegendProps) => {
  const styles = makeStyles(theme);
  return (
    <View style={styles.legend}>
      <View>
        <Text style={styles.legendTittle}>{title}</Text>
        <View style={styles.legendContentsWrapper}>
          {definition.map(item => (
            <View style={styles.legendContent}>
              <View style={item.style} />
              <Text style={[styles.legendText]}>{item.value}</Text>
            </View>
          ))}
        </View>
      </View>
    </View>
  );
};

interface TableHeadProps {
  theme: Theme;
  name: string;
  cellWidth: string;
  columns: string[];
}

const TableHead = ({ theme, name, cellWidth, columns }: TableHeadProps) => {
  const styles = makeStyles(theme);
  return (
    <View style={{ backgroundColor: theme.pdf?.table?.darkRowColor }}>
      <View>
        <Text style={styles.tableName}>{name}</Text>
      </View>
      <View style={[styles.tableHead, styles.tableRow]} wrap={false}>
        {columns.map(value => {
          return (
            <Text style={[styles.tableCell, { width: cellWidth }]}>
              {value}
            </Text>
          );
        })}
        <Text style={[styles.tableCell, { width: cellWidth }]}>Total</Text>
      </View>
    </View>
  );
};

interface TableProps {
  theme: Theme;
  name: string;
  rows: TableRowType[];
  columns: string[];
  cellWidth: string;
  showTotal: boolean;
  t: TFunction;
}

const Table = ({
  theme,
  name,
  rows,
  columns,
  cellWidth,
  showTotal,
  t,
}: TableProps) => {
  const styles = makeStyles(theme);

  const totals = showTotal
    ? columns.reduce<number[]>((colPrev, colCurr) => {
        const rowTotal = rows.reduce((rowPrev, rowCurr) => {
          const val = Number(rowCurr[colCurr]);
          if (!Number.isNaN(val)) {
            return rowPrev + val;
          }
          return rowPrev;
        }, 0);
        return [...colPrev, rowTotal];
      }, [])
    : [];

  const firstPageChunk = rows.slice(0, FIRST_PAGE_TABLE_ROWS);
  const restPagesChunks = chunk(
    rows.slice(FIRST_PAGE_TABLE_ROWS),
    MAX_TABLE_ROWS_PER_PAGE,
  );
  const chunks = [firstPageChunk, ...restPagesChunks];

  return (
    <>
      {chunks.map(chunkRow => {
        return (
          <View wrap={false}>
            <TableHead
              theme={theme}
              name={name}
              cellWidth={cellWidth}
              columns={columns}
            />
            {chunkRow.map((rowData, index) => {
              const color =
                index % 2
                  ? theme.pdf?.table?.darkRowColor
                  : theme.pdf?.table?.lightRowColor;
              const total = columns.reduce((prev, curr) => {
                const val = rowData[curr as keyof typeof rowData];
                if (!Number.isNaN(Number(val))) {
                  return prev + Number(val);
                }
                return prev;
              }, 0);
              return (
                <View
                  style={[styles.tableRow, { backgroundColor: color }]}
                  wrap={false}
                >
                  {columns.map(col => (
                    <Text style={[styles.tableCell, { width: cellWidth }]}>
                      {getTableCellVal(rowData, col, t)}
                    </Text>
                  ))}
                  <Text style={[styles.tableCell, { width: cellWidth }]}>
                    {Number(total).toLocaleString('en-US')}
                  </Text>
                </View>
              );
            })}
          </View>
        );
      })}

      {showTotal && (
        <View
          wrap={false}
          style={[
            styles.tableRow,
            styles.tableFooter,
            {
              backgroundColor:
                rows.length % 2
                  ? theme.pdf?.table?.darkRowColor
                  : theme.pdf?.table?.lightRowColor,
            },
          ]}
        >
          {totals.map((val, index) => {
            if (index === 0) {
              return (
                <Text style={[styles.tableCell, { width: cellWidth }]}>
                  Total
                </Text>
              );
            }
            return (
              <Text style={[styles.tableCell, { width: cellWidth }]}>
                {val}
              </Text>
            );
          })}
          <Text style={[styles.tableCell, { width: cellWidth }]}>
            {totals.reduce((prev, cur) => {
              return prev + cur;
            }, 0)}
          </Text>
        </View>
      )}
    </>
  );
};

const ReportDoc = ({
  theme,
  reportType,
  mapImage,
  tableData,
  tableName,
  tableRowsNum,
  tableShowTotal,
  eventName,
  sortByKey,
  exposureLegendDefinition,
  t,
}: ReportDocProps) => {
  if (mapImage === null) {
    return <Document />;
  }
  const styles = makeStyles(theme);
  const date = new Date().toUTCString();
  const tableCellWidth = `${100 / (tableData.columns.length + 1)}%`;

  const tableRows = tableData.rows.slice(1);
  const sortedTableRows =
    sortByKey !== undefined
      ? // eslint-disable-next-line fp/no-mutating-methods
        tableRows.sort((a, b) => {
          if (a[sortByKey] > b[sortByKey] || Number.isNaN(b[sortByKey])) {
            return -1;
          }
          if (a[sortByKey] < b[sortByKey] || Number.isNaN(a[sortByKey])) {
            return 1;
          }
          return 0;
        })
      : tableRows;
  const trimmedTableRows =
    tableRowsNum !== undefined
      ? sortedTableRows.slice(0, tableRowsNum - (tableShowTotal ? 1 : 0))
      : sortedTableRows;

  const areasLegendDefinition: PDFLegendDefinition[] = [
    {
      value: 'Province',
      style: [styles.dash, { backgroundColor: '#000000' }],
    },
    {
      value: 'District',
      style: [styles.dash, { backgroundColor: '#999797' }],
    },
    {
      value: 'Township',
      style: [styles.dash, { backgroundColor: '#D8D6D6' }],
    },
  ];

  const stormWindBuffersLegendDefinition: PDFLegendDefinition[] = [
    {
      value: 'Uncertainty Cones',
      style: [
        styles.borderedBox,
        { backgroundColor: '#ffffff', borderColor: '#b8b1b1' },
      ],
    },
    {
      value: 'Wind Buffer 60 km/h',
      style: [
        styles.borderedBox,
        { backgroundColor: '#fffcf1', borderColor: '#f7e705' },
      ],
    },
    {
      value: 'Wind Buffer 90 km/h',
      style: [
        styles.borderedBox,
        { backgroundColor: '#ffeed8', borderColor: '#f99408' },
      ],
    },
    {
      value: 'Wind Buffer 120 km/h',
      style: [
        styles.borderedBox,
        { backgroundColor: '#fcd4ce', borderColor: '#f90c08' },
      ],
    },
  ];

  const floodsLegendDefinition: PDFLegendDefinition[] = [
    {
      value: 'flooded',
      style: [styles.box, { backgroundColor: '#a50f15' }],
    },
  ];

  const populationExposureLegendDefinition: PDFLegendDefinition[] = exposureLegendDefinition.map(
    item => ({
      value: getLegendItemLabel(item),
      style: [styles.box, { backgroundColor: item.color, opacity: 0.5 }],
    }),
  );

  return (
    <Document>
      <Page size="A4" style={styles.page}>
        <View style={[styles.section]}>
          <Text style={styles.title}>Event name: {eventName}</Text>
          <Text style={styles.title}>Publication date: {date}</Text>
          <Text style={styles.subText}>
            This is an automated report.
            <Text> Information should be treated as preliminary</Text>
          </Text>
        </View>
        <View style={styles.section}>
          <Image src={mapImage} style={styles.mapImage} />
        </View>
        <View style={[styles.legendsContainer, styles.section]}>
          <Legend
            theme={theme}
            title="Areas"
            definition={areasLegendDefinition}
          />
          {reportType === ReportType.Storm ? (
            <Legend
              title="Tropical Storms - Wind buffers"
              definition={stormWindBuffersLegendDefinition}
              theme={theme}
            />
          ) : (
            <Legend
              title="Potential Flooding"
              definition={floodsLegendDefinition}
              theme={theme}
            />
          )}
          <Legend
            title="Population Exposure"
            definition={populationExposureLegendDefinition}
            theme={theme}
          />
        </View>
        <View style={styles.section}>
          <Text style={{ fontSize: theme.pdf?.fontSizes.small }}>
            Sources WFP, UNGIWG, OCHA, GAUL, USGS, NASA, UCSB
          </Text>
          <Text
            style={{
              fontSize: theme.pdf?.fontSizes.extraSmall,
              color: theme.pdf?.secondaryTextColor,
            }}
          >
            The designations employed and the presentation of material in the
            map(s) do not imply the expression of any opinion on the part of WFP
            concerning the legal or constitutional status of any country,
            territory, city or sea, or concerning the delimitation of its
            frontiers or boundaries.
          </Text>
        </View>
        <View style={[styles.section]}>
          <Table
            theme={theme}
            name={tableName}
            rows={trimmedTableRows}
            columns={tableData.columns}
            cellWidth={tableCellWidth}
            showTotal={tableShowTotal}
            t={t}
          />
        </View>
        <Text fixed style={styles.footer}>
          P R I S M automated report
        </Text>
      </Page>
    </Document>
  );
};

interface ReportDocProps {
  theme: Theme;
  reportType: ReportType;
  mapImage: string | null;
  tableData: TableData;
  tableName: string;
  tableRowsNum?: number;
  tableShowTotal: boolean;
  eventName: string;
  sortByKey?: keyof TableRowType;
  exposureLegendDefinition: LegendDefinition;
  t: TFunction;
}

export default ReportDoc;
