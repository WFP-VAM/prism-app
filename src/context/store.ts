import {
  combineReducers,
  configureStore,
  getDefaultMiddleware,
} from '@reduxjs/toolkit';

import mapStateReduce from './mapStateSlice';
import serverStateReduce from './serverStateSlice';
import tableStateReduce from './tableStateSlice';
import tooltipStateReduce from './tooltipStateSlice';
import notificationStateReduce, {
  errorToNotificationMiddleware,
} from './notificationStateSlice';
import analysisResultStateSlice from './analysisResultStateSlice';

const reducer = combineReducers({
  mapState: mapStateReduce,
  serverState: serverStateReduce,
  tableState: tableStateReduce,
  tooltipState: tooltipStateReduce,
  analysisResultState: analysisResultStateSlice,
  notificationState: notificationStateReduce,
});

export const store = configureStore({
  reducer,
  // TODO: Instead of snoozing this check, we might want to
  // serialize the state
  middleware: getDefaultMiddleware({
    serializableCheck: false,
    immutableCheck: {
      ignoredPaths: ['mapState.layersData', 'analysisResultState.result'],
    },
  }).concat(errorToNotificationMiddleware),
});

export type AppDispatch = typeof store.dispatch;
export type RootState = ReturnType<typeof reducer>;

export type ThunkApi = {
  dispatch: AppDispatch;
  getState: () => RootState;
  // can add 'extra' and 'rejectValue' here if we ever need them
  // https://redux-toolkit.js.org/usage/usage-with-typescript#createasyncthunk
};

// Used as the third type definition for typing createAsyncThunk
// e.g. createAsyncThunk<ReturnType, HandlerType, CreateAsyncThunkTypes>
export type CreateAsyncThunkTypes = {
  dispatch: AppDispatch;
  state: RootState;
};
