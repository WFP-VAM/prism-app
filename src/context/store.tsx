import { configureStore, getDefaultMiddleware } from '@reduxjs/toolkit';
import filtersReduce from './filters/filtersSlice';

export const store = configureStore({
  reducer: {
    filters: filtersReduce,
  },
  // TODO: Instead of snoozing this check, we might want to
  // serialize the state
  middleware: getDefaultMiddleware({
    serializableCheck: false,
  }),
});

export type RootState = ReturnType<typeof store.getState>;
