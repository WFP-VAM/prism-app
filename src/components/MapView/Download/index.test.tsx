import React from 'react';
import { render } from '@testing-library/react';

import { Provider } from 'react-redux';
import Download from '.';
import { store } from '../../../context/store';

test('renders as expected', () => {
  const { container } = render(
    <Provider store={store}>
      <Download />
    </Provider>,
  );
  expect(container).toMatchSnapshot();
});
