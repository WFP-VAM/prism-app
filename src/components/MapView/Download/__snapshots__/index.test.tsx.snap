// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders as expected 1`] = `
<div>
  <div
    class="MuiGrid-root MuiGrid-item"
  >
    <mock-button
      color="primary"
      variant="contained"
    >
      <mock-clouddownload
        fontsize="small"
      />
      <mock-hidden
        smdown="true"
      >
        <mock-typography
          classname="Download-label-1"
          variant="body2"
        >
          Export
        </mock-typography>
      </mock-hidden>
      <mock-arrowdropdown
        fontsize="small"
      />
    </mock-button>
    <mock-menu
      anchororigin="[object Object]"
      classes="[object Object]"
      elevation="0"
      id="export-menu"
      keepmounted="true"
      open="false"
      transformorigin="[object Object]"
    >
      <mock-menuitem
        classes="[object Object]"
      >
        <div
          class="MuiListItemIcon-root"
        >
          <mock-image
            fontsize="small"
            style="color: white;"
          />
        </div>
        <mock-listitemtext
          primary="IMAGE"
        />
      </mock-menuitem>
    </mock-menu>
    <mock-dialog
      aria-labelledby="dialog-preview"
      keepmounted="true"
      maxwidth="xl"
      open="false"
    >
      <mock-dialogtitle
        classname="Download-title-2"
        id="dialog-preview"
      >
        Map Preview
      </mock-dialogtitle>
      <mock-dialogcontent>
        <canvas />
      </mock-dialogcontent>
      <mock-dialogactions>
        <mock-button
          color="primary"
        >
          Cancel
        </mock-button>
        <mock-button
          color="primary"
          variant="contained"
        >
          Download PNG
        </mock-button>
        <mock-button
          color="primary"
          variant="contained"
        >
          Download JPEG
        </mock-button>
        <mock-button
          color="primary"
          variant="contained"
        >
          Download PDF
        </mock-button>
      </mock-dialogactions>
    </mock-dialog>
  </div>
</div>
`;
